pkg/capabilities/triggers/cron_trigger.go

```diff
@@ -0,0 +1,287 @@
+	TriggerID string `json:"triggerId"`
+	// The crontab to evaluate, with second granularity.
+	// The seconds field is required. The time zone will always be automatically set to UTC.
+	Schedule string `json:"schedule"`
+}
+
+type cronTriggerInput struct{}
+
+type CronTriggerPayload struct {
+	ScheduledExecutionTime string
+}
+
+type cronJob struct {
+	ch  chan<- capabilities.CapabilityResponse
+	job gocron.Job
+}
+
+type CronTriggerService struct {
+	capabilities.CapabilityInfo
+	capabilities.Validator[cronTriggerConfig, cronTriggerInput, capabilities.CapabilityResponse]
+	jobs      map[string]cronJob
+	lggr      logger.Logger
+	mu        sync.Mutex
+	scheduler gocron.Scheduler
+	stopCh    services.StopChan
+	wg        sync.WaitGroup
+}
+
+var _ capabilities.TriggerCapability = (*CronTriggerService)(nil)
+var _ services.Service = &CronTriggerService{}
+
+func NewCronTriggerService(lggr logger.Logger) *CronTriggerService {
+	s, err := gocron.NewScheduler()
+	if err != nil {
+		return nil
+	}
+	return &CronTriggerService{
+		CapabilityInfo: cronTriggerInfo,
+		Validator:      capabilities.NewValidator[cronTriggerConfig, cronTriggerInput, capabilities.CapabilityResponse](capabilities.ValidatorArgs{Info: cronTriggerInfo}),
+		jobs:           map[string]cronJob{},
+		lggr:           logger.Named(lggr, "CronTrigger"),
+		scheduler:      s,
+		stopCh:         make(services.StopChan),
+	}
+}
+
+func (cts *CronTriggerService) RegisterTrigger(ctx context.Context, req capabilities.CapabilityRequest) (<-chan capabilities.CapabilityResponse, error) {
+	cts.mu.Lock()
+	defer cts.mu.Unlock()
+
+	config, err := cts.ValidateConfig(req.Config)
+	if err != nil {
+		return nil, err
+	}
+
+	triggerID := getTriggerID(config.TriggerID, req.Metadata.WorkflowID)
+	if _, ok := cts.jobs[triggerID]; ok {
+		return nil, fmt.Errorf("triggerId %s already registered", triggerID)
+	}
+
+	job := gocron.CronJob(config.Schedule, true)
+
+	cronParser := cron.NewParser(
+		cron.SecondOptional | cron.Minute | cron.Hour | cron.Dom | cron.Month | cron.Dow,
+	)
+	schedule, err := cronParser.Parse(config.Schedule)
+	if err != nil {
+		return nil, err
+	}
+
+	t := gocron.NewTask(
+		cts.makeProcess(triggerID, schedule),
+	)
+
+	j, err := cts.scheduler.NewJob(
+		job,
+		t,
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	ch := make(chan capabilities.CapabilityResponse, defaultSendChannelBufferSize)
+	cts.jobs[triggerID] = cronJob{
+		ch:  ch,
+		job: j,
+	}
+	cts.lggr.Debugw("RegisterTrigger", "triggerId", triggerID, "jobId", j.ID())
+	return ch, nil
+}
+
+func (cts *CronTriggerService) makeProcess(triggerId string, schedule cron.Schedule) func() {
+	return func() { cts.process(triggerId, schedule) }
+}
+
+func (cts *CronTriggerService) process(triggerId string, schedule cron.Schedule) {
+	cts.mu.Lock()
+	defer cts.mu.Unlock()
+
+	now := time.Now()
+	scheduledExecutionTime := schedule.Next(now)
```

Comment:
You might be able to simplify this log by passing the job into the task closure and then calling j.LastRun()

Reply:
See here for an example: https://go.dev/play/p/EdyoEzZBihj

Reply:
Way nicer!